File: backend/services/prompt_builder.py
"""
Prompt builder service for constructing agent system prompts.

This module provides centralized prompt building logic to avoid duplication
across CRUD operations.
"""

import logging

from domain.entities.agent_config import AgentConfigData
from i18n.korean import format_with_particles
from sdk.loaders import get_base_system_prompt

logger = logging.getLogger(__name__)


def build_system_prompt(agent_name: str, config_data: AgentConfigData) -> str:
    """
    Build a complete system prompt for an agent.

    This function combines the base system prompt with agent-specific
    configuration markdown, applying Korean particle formatting.

    Uses different base prompts for different agent types:
    - Action Manager: system_prompt_AM
    - Onboarding Manager: system_prompt_OM
    - Other agents: active_system_prompt (default)

    Args:
        agent_name: The name of the agent
        config_data: Agent configuration data

    Returns:
        Complete system prompt string with markdown formatting
    """
    # Start with base system prompt (selected by agent type) and apply Korean particle formatting
    system_prompt = format_with_particles(get_base_system_prompt(agent_name), agent_name=agent_name)

    # Append character configuration with markdown headings
    config_markdown = config_data.to_system_prompt_markdown(agent_name)
    if config_markdown:
        system_prompt += config_markdown

    return system_prompt


def build_runtime_system_prompt(
    agent_name: str,
    config_data: AgentConfigData,
    lore: str | None = None,
) -> str:
    """
    Build a system prompt at runtime with optional lore injection.

    Structure:
        [platform_guideline]
        [lore]                 <- injected between guideline and traits
        [character_config]

    This allows world-specific lore to be injected between the platform
    guidelines and character traits for better context.

    Uses different base prompts for different agent types:
    - Action Manager: system_prompt_AM
    - Onboarding Manager: system_prompt_OM
    - Other agents: active_system_prompt (default)

    Args:
        agent_name: The name of the agent
        config_data: Agent configuration data
        lore: Optional world lore to inject

    Returns:
        Complete system prompt with lore injected in the middle
    """
    # Start with base system prompt (selected by agent type) and apply Korean particle formatting
    system_prompt = format_with_particles(get_base_system_prompt(agent_name), agent_name=agent_name)

    # Inject lore between platform guideline and character config
    if lore:
        system_prompt += f"\n\n# World Lore\n\n{lore.strip()}"

    # Append character configuration with markdown headings
    config_markdown = config_data.to_system_prompt_markdown(agent_name)
    if config_markdown:
        system_prompt += config_markdown

    return system_prompt

File: backend/src/services/prompt-builder.ts
/**
 * System prompt assembly. The character sheet is appended as markdown headings
 * rather than XML tags, so it blends into the SDK's own immutable "You are
 * Claude Code" prompt instead of fighting it. Section order is load-bearing:
 * identity, behaviour, recent context, memory index — cheapest-to-ignore
 * material furthest from the instructions.
 */

import { formatWithParticles } from '../lib/korean'
import { getBaseSystemPrompt } from '../sdk/loaders/guidelines'
import type { AgentConfigData } from '../sdk/parsing/agent-config'

export type PromptConfigData = Pick<
  AgentConfigData,
  'inANutshell' | 'characteristics' | 'recentEvents' | 'longTermMemorySubtitles'
>

/**
 * Render the character sheet as markdown, or `''`. The leading blank lines are
 * part of the contract — callers concatenate this onto the base prompt. The
 * memory heading's hardcoded `이` particle deliberately skips
 * `formatWithParticles`: fixing it would change every agent's prompt.
 */
export function toSystemPromptMarkdown(agentName: string, config: PromptConfigData): string {
  const sections: string[] = []

  if (config.inANutshell) {
    sections.push(`## ${agentName} in a nutshell\n\n${config.inANutshell}`)
  }
  if (config.characteristics) {
    sections.push(`## ${agentName}'s characteristics\n\n${config.characteristics}`)
  }
  if (config.recentEvents) {
    sections.push(`## ${agentName}'s recent events\n\n${config.recentEvents}`)
  }
  if (config.longTermMemorySubtitles) {
    sections.push(`## ${agentName}이 가진 기억 index\n\n${config.longTermMemorySubtitles}`)
  }

  return sections.length > 0 ? `\n\n${sections.join('\n\n')}` : ''
}

function baseFor(agentName: string): string {
  return formatWithParticles(getBaseSystemPrompt(agentName), { agent_name: agentName })
}

// The base prompt varies by agent type; Action Manager and Onboarding Manager
// get dedicated ones.
export function buildSystemPrompt(agentName: string, config: PromptConfigData): string {
  return baseFor(agentName) + toSystemPromptMarkdown(agentName, config)
}

/**
 * Build a system prompt at request time, optionally injecting world lore.
 * Guidelines → lore → character sheet: lore sits in the middle so the traits
 * stay adjacent to the agent's own output, where they have most pull.
 */
export function buildRuntimeSystemPrompt(
  agentName: string,
  config: PromptConfigData,
  lore?: string | null,
): string {
  let prompt = baseFor(agentName)
  if (lore) prompt += `\n\n# World Lore\n\n${lore.trim()}`
  return prompt + toSystemPromptMarkdown(agentName, config)
}
